contracts/0.8.9/StakingRouter.sol

```diff
@@ -0,0 +1,525 @@
+    event WithdrawalCredentialsSet(bytes32 withdrawalCredentials);
+    event ContractVersionSet(uint256 version);
+    /**
+     * Emitted when the vault received ETH
+     */
+    event ETHReceived(uint256 amount);
+
+    struct StakingModule {
+        /// @notice unique id of the module
+        uint24 id;
+        /// @notice name of module
+        string name;
+        /// @notice address of module
+        address stakingModuleAddress;
+        /// @notice rewarf fee of the module
+        uint16 moduleFee;
+        /// @notice treasury fee
+        uint16 treasuryFee;
+        /// @notice target percent of total keys in protocol, in BP
+        uint16 targetShare;
+        /// @notice flag if module can not accept the deposits
+        bool paused;
+        /// @notice flag if module can participate in further reward distribution
+        bool active;
+        /// @notice block.timestamp of the last deposit of the module
+        uint64 lastDepositAt;
+        /// @notice block.number of the last deposit of the module
+        uint256 lastDepositBlock;
+    }
+
+    struct StakingModuleCache {
+        bool paused;
+        uint16 targetShare;
+        uint256 totalKeysCount;
+        uint256 usedKeysCount;
+        uint256 stoppedKeysCount;
+        uint256 activeKeysCount;
+        uint256 availableKeysCount;
+    }
+
+    ILido public immutable LIDO;
+
+    bytes32 public constant MANAGE_WITHDRAWAL_CREDENTIALS_ROLE = keccak256("MANAGE_WITHDRAWAL_CREDENTIALS_ROLE");
+    bytes32 public constant MODULE_PAUSE_ROLE = keccak256("MODULE_PAUSE_ROLE");
+    bytes32 public constant MODULE_RESUME_ROLE = keccak256("MODULE_RESUME_ROLE");
+    bytes32 public constant MODULE_MANAGE_ROLE = keccak256("MODULE_MANAGE_ROLE");
+    bytes32 public constant STAKING_ROUTER_DEPOSIT_ROLE = keccak256("STAKING_ROUTER_DEPOSIT_ROLE");
+
+    /// Version of the initialized contract data
+    /// NB: Contract versioning starts from 1.
+    /// The version stored in CONTRACT_VERSION_POSITION equals to
+    /// - 0 right after deployment when no initializer is invoked yet
+    /// - N after calling initialize() during deployment from scratch, where N is the current contract version
+    /// - N after upgrading contract from the previous version (after calling finalize_vN())
+    bytes32 internal constant CONTRACT_VERSION_POSITION = keccak256("lido.StakingRouter.contractVersion");
+
+    /// @dev Credentials which allows the DAO to withdraw Ether on the 2.0 side
+    bytes32 internal constant WITHDRAWAL_CREDENTIALS_POSITION = keccak256("lido.StakingRouter.withdrawalCredentials");
+
+    uint256 public constant DEPOSIT_SIZE = 32 ether;
+
+    uint256 internal constant TOTAL_BASIS_POINTS = 10000;
+
+    /// @dev total count of staking modules
+    uint256 private _stakingModulesCount;
+
+    /// @dev id of the last added staking module. This counter grow on staking modules adding
+    uint24 private _lastStakingModuleId;
```

Comment:
@Psirex why not use unstructed storage here?

Reply:
Move to unstructured storage